[#3] 로그인/로그아웃 기능 구현 PR

msugo1 opened the pull request:
## [#3] 로그인 및 로그아웃 기능 구현
- 유저의 서비스 이용을 위한 로그인 및 로그아웃 기능 구현(세션 방식)
- 로그인 성공/실패(예외)에 대한 테스트 케이스 작성
- 서비스계층 테스트 역할을 좀더 명확하게 분리(회원가입 요청 처리  테스트, 회원가입 시 유효성 검사 테스트, 유저 로그인 및 로그아웃 테스트)
- `Mockito`를 이용해 `UserController`를 다른 계층으로부터 고립시켜 단위테스트생성

Commit:
[#3] 로그인 및 로그아웃 기능 구현

- 유저의 서비스 이용을 위한 로그인 및 로그아웃 기능 구현(세션 방식)
- 로그인 성공/실패(예외)에 대한 테스트 케이스 작성
- 서비스계층 테스트 역할을 좀더 명확하게 분리(회원가입 요청 처리  테스트,
  회원가입 시 유효성 검사 테스트, 유저 로그인 및 로그아웃 테스트)

Commit:
Merge branch 'feature/2' into feature/3

* 'feature/2' 브랜치에서 반영한 업데이트 내역을 'feature/3' 브랜치로
  가져오기 위해 두 브랜치를 merge

Commit:
[#3] UserController 단위테스트 추가 및 기존 테스트 내용 수정/추가

- Mockito를 이용해 UserController를 다른 계층으로부터 고립시켜
  단위테스트생성
- 통합테스트 내 테스트 추가(로그인/로그아웃 성공, 실패 케이스 테스트)
- Mockito를 이용한 유저 회원가입 단위 테스트 업데이트

Commit:
[#3] Merge branch 'feature/2' into feature/3

* feature/2에서의 피드백 수정사항 반영하기 위한 merge

Commit:
[#3] 로그인 관련 메소드 내용 수정

* USerServiceImpl
- 이전 loginRequest 메소드는 올바른 로그인 정보를 입력했음에도 'IllegalArgumentException' 예외가 발생해 세션에 해당 값을 담지 못하는 경우가 발생함
(원인: DB내에 encode 된 암호비교가 제대로 이루어지지 않았음)
- DB내에서 패스워드 일치 여부를 체크하던 방식에서 UserLogin 객체를
  받아와 Service 계층에서 암호화된 패스워드의 일치여부를 확인하도록 변경

* UserRepository / UserMapper
- 로그인 요청이 들어오면 DB에서 해당하는 이름에 대한 데이터를 찾아
  아이디와 패스워드를 담은 UserLogin객체를 만들어
  리턴해주는'getRegisteredUserInfo' 메소드 추가

* Test
- 메소드 내용 변경에 따른 테스트 수정(통합 테스트, 단위 테스트 모두)
- DAO layer에서 DB를 연결해getRegisteredUserInfo가 제대로 작동하는지
  확인하는 테스트 추가

f-lab-dev commented on src/main/java/me/soo/helloworld/HelloworldApplication.java:
```diff
@@ -2,7 +2,9 @@
 import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.cache.annotation.EnableCaching;
 
+@EnableCaching
```
이 어노테이션은 어떤 것 때문에 달아주셨나요?

msugo1 replied:
`Reddis` 설정보고 따라하면서 붙여놨었던 것 같아요. 일단 이 단계에서는 Reddis가 필요하지 않아서 의존성이랑 `ReddisConfig` 파일은 제거 했는데 그 때 저 어노테이션은 잘 안보이는 위치에 있어서 깜빡한 것 같습니다

f-lab-dev commented on src/main/java/me/soo/helloworld/service/UserServiceImpl.java:
```diff
@@ -36,4 +40,30 @@ public void userSignUp(User user) {
+
+    @Override
+    public void loginRequest(UserLoginInfo userLoginInfo, HttpSession httpSession) {
+        if (httpSession.getAttribute("userId") != null) {
```
`httpSession.getAttribute("userId")` 이런 로그인(인증)과 관련된 책임은 별도로 빼는게 낫지 않을까요?

msugo1 replied:
넵 안그래도 작성해놓고 리팩토링이 필요해보여서 저 부분만 따로 분리해서 UserService에 다른 메소드로 만들어 놓는게 좋을지 아니면 로그인 인증관련 클래스 자체를 하나 더 만드는게 좋을지 계속 고민하고 있었습니다.  이 둘 중 선택에 근거가 되는 것이 차후에 다른 클래스/메소드에서도 같은 내용이 사용될지 인가요??

f-lab-dev replied:
단순 그것뿐만은 아니라 `단일 책임 원칙`도 생각해보시면 좋을 것 같아요.

**유저의 정보를 관리하는 것과 서비스의 인증/인가는 같은 책임인가** 에 대해서요

f-lab-dev commented on src/main/java/me/soo/helloworld/service/UserServiceImpl.java:
```diff
@@ -36,4 +40,30 @@ public void userSignUp(User user) {
+    }
+
+    @Override
+    public void logoutRequest(HttpSession httpSession) {
```
`Request`는 굳이 달지 않아도 로그아웃이라는 행위를 충분히 표현할 수 있을 것 같네요~

msugo1 replied:
네 `logout`으로 수정하겠습니다.

Commit:
[#3] 피드백 반영

* UserService 리팩토링
- 서비스 인증/인가에 대한 책임을 `LoginService`로 분리
- 비밀번호 인코딩에 따른 객체 변환/반환에 대한 책임을 `User`로
  이전

* SessionKeys 클래스 추가 (in util)
- 기존의 세션 키값이 `userId`로 각 코드마다 직접적으로  하드코딩 되어있던 부분을
  상수로 대체하여 코드중복 제거

* Test 수정
- UserService의 리팩토링에 뒤따라 `LoginService`관련 테스트 수정 및
  테스트 완료

f-lab-dev commented on src/main/java/me/soo/helloworld/service/LoginService.java:
```diff
@@ -0,0 +1,12 @@
+
+    public void login(UserLoginInfo requestedUserLoginInfo, UserLoginInfo storedUserLoginInfo, HttpSession httpSession);
+
+    public void logout(HttpSession httpSession);
```
만약 세션 방식이 아닌 로그인을 구현하려고 할때 이 `HttpSession` 인자가 문제가 되지 않을까요?

msugo1 replied:
로그인 인터페이스를 구현해놓은 이유가 나중에 로그인 방식을 바꿀 때 유연하게 대처하려고 한 부분인데 Session인자 자체가 이를 막는다는 것을 간과했네요. 

Service 계층 자체를 Session에서 자유롭게 만들어 줄 필요가 있다고 생각해서 이렇게 처리했었는데 그렇다면 그냥 Session을 Service 계층에서 주입받아서 사용하는 것도 하나의 해결책이 될 수 있을까요?

f-lab-dev replied:
네 `HttpSession`은 외부에서 주입받을 수 있습니다~

Commit:
[#3] 피드백 반영

* LoginService
- 세션 로그인 방식에 종속 되있던 메소드: HttpSession 파라미터 제거(Both
  Login and Logout)
- UserController에서 LoginService 계층으로 세션을 파라미터로 전달하는 대신 LoginService 계층에서 직접  주입해서 사용
- Login 메소드 내용 수정: 아이디의 null값 검증 절차를 UserController로
  이전

* UserController
- userLogin method 수정: 전달받은 'UserIdAndPassword' 객체가null인지
  체크하고 null일 경우 Status 코드 404(Not Found)를 리턴

* UserRepository, Usermapper
- getRegisteredInfo에서 getRegisteredUserById로 메소드 이름 변경
- 아이디와 패스워드가 담겨 있는 객체를 전달하는 것 대신에 userId만
  문자열 파라미터로 전달

* 테스트 관련
- 수정 내용과 관련된 테스트 내용 수정
- 실패하는 테스트는 차후 수정요망

f-lab-dev commented on src/main/java/me/soo/helloworld/model/user/UserIdAndPassword.java:
```diff
@@ -9,10 +9,12 @@
 @Getter
 @AllArgsConstructor
-public class UserLoginInfo {
+public class UserIdAndPassword {
```
이 클래스는 로그인 요청정보를 담는 클래스이므로 `LoginRequest`가 낫지 않을까요?

msugo1 replied:
변경했습니다! `LoginRequest` 바꿨더니 수정할 부분들이 좀 더 명확해지더라구요 감사합니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/repository/UserRepository.java:
```diff
@@ -20,7 +20,7 @@ public boolean isUserIdDuplicate(String userId) {
 
-    public UserLoginInfo getRegisteredUserInfo(UserLoginInfo userLoginInfo) {
-        return userMapper.getRegisteredUserInfo(userLoginInfo);
+    public UserIdAndPassword getRegisteredUserById(String userId) {
```
`RegisteredUser`가 결국 `User`아닐까요? `Registered`는 빼주셔도 될 것 같습니다~

msugo1 replied:
`getUserById`로 변경했습니다.

Commit:
[#3] 피드백 반영 및 UserService, SessionLoginService 클래스 수정

* 피드백 반영
- UserIdAndPassword -> LoginRequest로 변경
- UserMapper 메소드 수정: getRegisteredUserById -> getUserById

* SessionLoginService
- 아이디/비밀번호 확인 책임을 UserService로 이관. Sessino관련 인증만
  담당하도록 변경
- login메소드 파라미터로 userId(String)만 넘기도록 변경
- passwordEncoder 인스턴스 변수가 redundant 해졌으므로 제거

* UserService
- 좀더 범용적으로 사용될 가능성이 있는 메소드 이름 변경: getLoginUser -> getUser
- 기존에 `SessionLogin`에서 담당했던 유저의 존재 유무와 비밀번호 일치
  여부의 책임을임시적으로 `getUser`가 담당하도록 변경
  (올바른 ID와 비밀번호를 판단하는 책임을 어떻게 해야 getUser로부터 잘
  분리해낼 수 있을까...)

* UserRepository/UserMapper
- 로그인 정보가 아닌 로그인 정보에 해당하는 실제 유저객체를 리턴하도록
  getUserById 메소드 변경

* Test 관련
- 위의 변경사항에 영향을 받는 테스트 내용 알맞게 수정
- 불필요해 보이는 테스트파일들 삭제
- UserLogin, UserService관련 단위테스트 올바르게 재작성
